bundles/org.openhab.core.voice/src/main/java/org/openhab/core/voice/internal/VoiceManagerImpl.java

```diff
@@ -951,8 +1039,92 @@ private Comparator<Voice> createVoiceComparator(Locale locale) {
         return null;
     }
 
+    private void stopDialogs(Predicate<DialogProcessor> filter) {
+        synchronized (dialogRegistrationStorage) {
+            var dialogsToStop = dialogProcessors.values().stream().filter(filter).toList();
+            if (dialogsToStop.isEmpty()) {
+                return;
+            }
+            for (var dialog : dialogsToStop) {
+                stopDialog(dialog.dialogContext.source());
+            }
+        }
+    }
+
+    private void debounceBuildDialogRegistrations() {
+        if (dialogRegistrationFuture != null) {
+            dialogRegistrationFuture.cancel(false);
+        }
+        dialogRegistrationFuture = scheduledExecutorService.schedule(this::buildDialogRegistrations, 5,
+                TimeUnit.SECONDS);
+    }
+
+    private void buildDialogRegistrations() {
+        synchronized (dialogRegistrationStorage) {
+            dialogRegistrationStorage.getValues().stream().forEach(dr -> {
+                if (dr != null && !dialogProcessors.containsKey(dr.sourceId)) {
+                    this.tryBuildDialogRegistration(dr);
+                }
+            });
+        }
+    }
+
+    private void tryBuildDialogRegistration(DialogRegistration dialogRegistration) {
+        var builder = getDialogContextBuilder().withSink(audioManager.getSink(dialogRegistration.sinkId))
+                .withSource(audioManager.getSource(dialogRegistration.sourceId));
+        String ksId = dialogRegistration.ksId;
+        if (ksId != null) {
+            builder.withKS(getKS(ksId));
+        }
+        String keyword = dialogRegistration.keyword;
```

Comment:
You do a lot of these null-checks in several places. IMO it would be better to accept `null` in the builder and discard the value there. You can also improve `getKS` and similar:

```
    @Override
    public @Nullable KSService getKS(@Nullable String id) {
        return id == null ? null : ksServices.get(id);
    }
```

then you can really make good use of the builder and concatenate the `.withXY` calls.

Reply:
I don't like it either, but I didn't know how to solved.
I think I have understand your solution looks great, I'll go for it.

Reply:
Looks much better now, thank you for proposing this change.
I can take advantage of it also on this one https://github.com/openhab/openhab-core/pull/3253.